File: src/etch/bindings/java/transport/fmt/binary/BinaryTaggedDataInput.java
/* $Id$
 *
 * Created by sccomer on Jan 17, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.bindings.java.transport.fmt.binary;

import java.io.IOException;

import metreos.util.FlexBuffer;
import metreos.util.URL;
import etch.bindings.java.msg.ArrayValue;
import etch.bindings.java.msg.Message;
import etch.bindings.java.msg.StructValue;
import etch.bindings.java.msg.TaggedDataInput;
import etch.bindings.java.msg.Type;
import etch.bindings.java.msg.Validator;
import etch.bindings.java.msg.Validator_int;
import etch.bindings.java.msg.ValueFactory;
import etch.bindings.java.support.MessageSource;
import etch.bindings.java.transport.fmt.TypeCode;

/**
 * BinaryTaggedDataInput has methods to support reading tagged
 * values from an input buffer.
 */
final public class BinaryTaggedDataInput extends BinaryTaggedData
	implements TaggedDataInput
{
	/**
	 * Constructs the BinaryTaggedDataInput.
	 * 
	 * @param vf the value factory for the service.
	 * @param buf the input buffer.
	 */
	public BinaryTaggedDataInput( ValueFactory vf, FlexBuffer buf )
	{
		super( vf );
		this.buf = buf;
	}
	
	/**
	 * Constructs the BinaryTaggedDataInput with a null buffer.
	 * 
	 * @param vf the value factory for the service.
	 * @param uri the uri used to construct the transport stack.
	 * 
	 * @see #setBuffer(FlexBuffer)
	 */
	public BinaryTaggedDataInput( ValueFactory vf, URL uri )
	{
		super( vf );
		// don't have anything to do with uri yet.
	}

	private FlexBuffer buf;

	/**
	 * @param buf
	 */
	public void setBuffer( FlexBuffer buf )
	{
		this.buf = buf;
	}

	public Message startMessage() throws IOException
	{
		byte version = buf.getByte();
		
		if (version != VERSION)
			throw new IOException(
				String.format(
					"binary tagged data version mismatch: got %d expected %d",
					version, VERSION ) );
		
		Type type = readStructType();
		return new Message( type, vf );
	}

	public void endMessage( Message msg )
	{
		// nothing to do.
	}

	public StructValue startStruct() throws IOException
	{
		Type type = readStructType();
		return new StructValue( type );
	}
	
	private Type readStructType() throws IOException
	{
		Integer id = (Integer) readValue( intValidator );
		return vf.getType( id );
	}

	public boolean readStructElement( StructElement se ) throws IOException
	{
		Object obj = readValueOrNone( intValidator );
		if (obj == NONE)
			return false;
		
		Integer id = (Integer) obj;
		se.key = vf.getField( id );
		
		Validator v = currentType != null ? currentType.getValidator( se.key ) : null;
		se.value = readValue( v );
		return true;
	}
	
	private Type currentType;

	public void endStruct( StructValue struct )
	{
		// nothing to do.
	}

	public ArrayValue startArray()
		throws IOException
	{
		byte type = buf.getByte();
		
		Type customStructType;
		if (type == TypeCode.CUSTOM)
			customStructType = readStructType();
		else
			customStructType = null;
		
		int dim = (Integer) readValue( intValidator );
		
//		System.out.printf( "startArray (input) %d %s %d\n", type, customStructType, dim );
		
		return new ArrayValue( type, customStructType, dim );
	}

	public boolean readArrayElement( ArrayElement ae )
		throws IOException
	{
		ae.value = readValueOrNone( arrayElementValidator );
		if (ae.value == NONE)
			return false;
		return true;
	}
	
	private Validator arrayElementValidator;

	public void endArray( ArrayValue array )
	{
		// nothing to do.
	}

	/**
	 * Reads a Message from the data buffer.
	 * @param ms the message source.
	 * @param vf the value factor for the message.
	 * @param buf the data buffer.
	 * @return the message read.
	 * @throws IOException if there is a problem reading the message.
	 */
	public static Message readMessage( ValueFactory vf, byte[] buf,
		MessageSource ms ) throws IOException
	{
		return readMessage( vf, new FlexBuffer( buf ), ms );
	}
	
	/**
	 * Reads a Message from the data buffer.
	 * @param ms the message source.
	 * @param vf the value factor for the message.
	 * @param buf the data buffer.
	 * @param off offset to start of message.
	 * @param len length of message.
	 * @return the message read.
	 * @throws IOException if there is a problem reading the message.
	 */
	public static Message readMessage( ValueFactory vf, byte[] buf, int off,
		int len, MessageSource ms ) throws IOException
	{
		return readMessage( vf, new FlexBuffer( buf, off, len ), ms );
	}
	
	/**
	 * @param ms
	 * @param vf
	 * @param buf
	 * @return the message read.
	 * @throws IOException
	 */
	public static Message readMessage( ValueFactory vf, FlexBuffer buf,
		MessageSource ms ) throws IOException
	{
		BinaryTaggedDataInput tdi = new BinaryTaggedDataInput( vf, buf );
		return tdi.readMessage();
	}
	
	public Message readMessage() throws IOException
	{
		Message msg = startMessage();
		readKeysAndValues( msg );
		endMessage( msg );
		return msg;
	}
	
//	/**
//	 * @return a signed byte (-128..127) from the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private byte readByte() throws IOException
//	{
//		int b = is.read();
//		// is.read() returns 0-255 value or -1 if eof
//		if (b < 0)
//			throw new EOFException();
//		return (byte) b;
//	}

//	/**
//	 * @return an unsigned byte (0..255) from the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private short readUByte() throws IOException
//	{
//		int b = is.read();
//		// is.read() returns 0-255 value or -1 if eof
//		if (b < 0)
//			throw new EOFException();
//		return (short) b;
//	}
	
//	/**
//	 * @return a signed short from the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private short readShort() throws IOException
//	{
//		short a = readUByte();
//		short b = readUByte();
//		return (short) (a | (b << 8));
//	}
	
//	/**
//	 * @return an unsigned short from the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private int readUShort() throws IOException
//	{
//		int a = readUByte();
//		int b = readUByte();
//		return a | (b << 8);
//	}

//	/**
//	 * @return a signed int from the input stream.
//	 * @throws IOException
//	 */
//	public int readInt() throws IOException
//	{
//		int a = readUShort();
//		int b = readUShort();
//		return a | (b << 16);
//	}
	
//	/**
//	 * @return an unsigned int from the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private long readUInt() throws IOException
//	{
//		long a = readUShort();
//		long b = readUShort();
//		return a | (b << 16);
//	}
	
//	/**
//	 * @return a signed long from the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private long readLong() throws IOException
//	{
//		long a = readUInt();
//		long b = readUInt();
//		return a | (b << 32);
//	}
	
//	/**
//	 * @return an IEEE format single precision floating point number from
//	 * the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private float readFloat() throws IOException
//	{
//		return Float.intBitsToFloat( readInt() );
//	}
	
//	/**
//	 * @return an IEEE format double precision floating point number from
//	 * the input stream.
//	 * @throws IOException if there is a problem with the input stream.
//	 */
//	private double readDouble() throws IOException
//	{
//		return Double.longBitsToDouble( readLong() );
//	}
	
//	/**
//	 * Read all the bytes from the input stream.
//	 * @param b a buffer for the bytes that are read.
//	 * @throws IOException if there is a problem with the input stream. 
//	 */
//	private void readBytes( byte[] b ) throws IOException
//	{
//		int n = b.length;
//		for (int i = 0; i < n; i++)
//			b[i] = readByte();
//	}

//	private byte readType() throws IOException
//	{
//		return readByte();
//	}
	
	/**
	 * @param v the validator to use to check the value format.
	 * @return the object read.
	 * @throws IOException
	 */
	private Object readValue( Validator v )
		throws IOException
	{
		Object value = readValue0();
		
		if (value == null)
			return null;
		
		return validateValue( value, v );
	}

	private Object readValueOrNone( Validator v )
		throws IOException
	{
		Object value = readValue0();
		
		if (value == null)
			return null;
		
		if (value == NONE)
			return value;
		
		return validateValue( value, v );
	}
	
	private Object validateValue( Object value, Validator v )
	{
		// TODO replace with if (v == null || !v.validate( value ))
		if (v != null && !v.validate( value ))
		{
			// issue a warning?
			System.out.printf( "value %s does not validate using %s\n", value, v );
			return null;
		}
		
		return value;
	}
	
	/**
	 * @return the object read.
	 * @throws IOException
	 */
	private Object readValue0()
		throws IOException
	{
		byte type = buf.getByte();
		
		switch (type)
		{
			case TypeCode.NULL:
				return null;
			
			case TypeCode.BOOLEAN_FALSE:
				return Boolean.FALSE;
			
			case TypeCode.BOOLEAN_TRUE:
				return Boolean.TRUE;
			
			case TypeCode.BYTE1:
				return buf.getByte();
			
			case TypeCode.SHORT1:
				return (short) buf.getByte();
			
			case TypeCode.INT1:
				return (int) buf.getByte();
			
			case TypeCode.LONG1:
				return (long) buf.getByte();
			
			case TypeCode.SHORT2:
				return buf.getShort();
			
			case TypeCode.INT2:
				return (int) buf.getShort();
			
			case TypeCode.LONG2:
				return (long) buf.getShort();
			
			case TypeCode.INT4:
				return buf.getInt();
			
			case TypeCode.LONG4:
				return (long) buf.getInt();
			
			case TypeCode.LONG8:
				return buf.getLong();
			
			case TypeCode.FLOAT4:
				return buf.getFloat();
			
			case TypeCode.FLOAT8:
				return buf.getDouble();
				
			case TypeCode.BYTES:
				return readBytes();
			
			case TypeCode.EMPTY_STRING:
				return "";
			
			case TypeCode.STRING:
				return new String( readBytes(), vf.getStringEncoding() );
			
			case TypeCode.ARRAY:
				return fromArrayValue( readArrayValue() );
			
			case TypeCode.STRUCT:
				return readStructValue();
			
			case TypeCode.CUSTOM:
				return vf.importCustomValue( readStructValue() );
			
			case TypeCode.NONE:
				return NONE;
			
			default:
				if ((type & TypeCode.PSMASK) == TypeCode.PSVALUE)
				{
					// positive small integers
					int value = type & TypeCode.PVMASK;
					int dt = (type >> TypeCode.PDTSHIFT) & TypeCode.DTMASK;
					if (dt == TypeCode.BYTE_DT) return (byte) value;
					if (dt == TypeCode.SHORT_DT) return (short) value;
					if (dt == TypeCode.INT_DT) return value;
					return (long) value;
				}
				if ((type & TypeCode.NSMASK) == TypeCode.NSVALUE)
				{
					// negative small integers
					int value = -(type & TypeCode.NVMASK)-1;
					int dt = (type >> TypeCode.NDTSHIFT) & TypeCode.DTMASK;
					if (dt == TypeCode.BYTE_DT) return (byte) value;
					if (dt == TypeCode.SHORT_DT) return (short) value;
					if (dt == TypeCode.INT_DT) return value;
					return (long) value;
				}
				throw new UnsupportedOperationException( "unsupported type code "+type );
		}
	}
	
	public ArrayValue readArrayValue() throws IOException
	{
		ArrayValue array = startArray();
		readValues( array );
		endArray( array );
		return array;
	}

	private void readValues( ArrayValue array ) throws IOException
	{
		ArrayElement ae = new ArrayElement();
		while (readArrayElement( ae ))
			array.add( ae.value );
	}

	public StructValue readStructValue() throws IOException
	{
		StructValue sv = startStruct();
		readKeysAndValues( sv );
		endStruct( sv );
		return sv;
	}

	/**
	 * Reads the key / value pairs and puts them in the struct value.
	 * @param sv
	 * @throws IOException if there are problems with the tagged data input.
	 */
	private void readKeysAndValues( StructValue sv )
		throws IOException
	{
		StructElement se = new StructElement();
		while (readStructElement( se ))
			sv.put( se.key, se.value );
	}
	
	private byte[] readBytes() throws IOException
	{
		int n = (Integer) readValue( intValidator );
		byte[] b = new byte[n];
		buf.getFully( b );
		return b;
	}
	
	private final Validator intValidator = Validator_int.get( 0 );
}

File: src/etch/bindings/csharp/transport/fmt/binary/BinaryTaggedDataInput.cs
using System;
using System.IO;
using System.Text;
using System.Collections.Generic;
using Etch.Msg;
using Etch.Util;
using Etch.Support;

namespace Etch.Transport.Fmt.Binary
{
    /// <summary>
    /// TaggedDataInputStream has methods to support reading little-endian integer
    /// and floating point data types as well as tagged values from an input
    /// stream.
    /// </summary>
    public class BinaryTaggedDataInput : BinaryTaggedData, TaggedDataInput
    {
        /// <summary>
        /// Constructs the BinaryTaggedDataInput.
        /// </summary>
        /// <param name="vf">the value factory for the service</param>
        /// 
        public BinaryTaggedDataInput(ValueFactory vf) : base(vf)
        {
            
          
        }

        /// <summary>
        /// Constructs the BinaryTaggedDataInput with a null buffer.
        /// </summary>
        /// <param name="vf">the value factory for the service.</param>
        /// <param name="uri">the uri used to construct the transport stack.</param>
        public BinaryTaggedDataInput( ValueFactory vf, string uri ) : base( vf )
        {}

        private FlexBuffer buf;

        /////////////////////////////
        // TaggedDataInput methods //
        /////////////////////////////


        public Message ReadMessage(FlexBuffer buf)
        {
            this.buf = buf;
            try
            {
                Message msg = StartMessage();
                ReadKeysAndValues(msg);
                EndMessage(msg);
                return msg;
            }
            finally
            {
                this.buf = null;
            }

        }

        private StructValue ReadStruct()
        {
            StructValue sv = StartStruct();
            ReadKeysAndValues( sv );
            EndStruct( sv );
            return sv;

        }

        private ArrayValue ReadArray( Validator v )
        {
            ArrayValue av = StartArray();
            ReadValues( av, v );
            EndArray( av );
            return av;
        }


        private void ReadKeysAndValues( StructValue sv )
        {
            XType t = sv.GetXType;

            while ( true )
            {
                Object obj = ReadValue( intValidator, true );
                if ( obj == NONE )
                    break;

                int? id = ( int? ) obj;
                Field key = t.GetField( id );

                Validator v = t.GetValidator( key );
                Object value = ReadValue( v );

                sv.Add( key, value );
            }
        }

        private void ReadValues( ArrayValue av, Validator v )
        {
            Validator ev = v.ElementValidator();
            while ( true )
            {
                Object value = ReadValue( ev, true );
                if ( value == NONE )
                    break;

                av.Add( value );
            }
        }

        ////////////////////////
        // Main input methods //
        ////////////////////////

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        /// Exception:
        ///     throws Exception
        private Message StartMessage()
        {
            sbyte version = ( sbyte ) buf.GetByte();
			
			if (version != VERSION)
				throw new IOException(
					"binary tagged data version mismatch: got "+version+" expected "+VERSION );

            XType t = ReadType();
            int length = ReadIntegerValue().Value;

      //      return new Message( ReadType(), vf );
            return new Message(t,length, vf);
        }

        private void EndMessage( Message msg )
        {
            // nothing to do., readKeysAndValues swallowed the NONE.
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        /// Exception:
        ///     throws Exception
        private StructValue StartStruct() 
	    {
            XType t = ReadType();
            int length = ReadIntegerValue().Value;
            return new StructValue( t,length);
	    }

        private void EndStruct( StructValue sv )
        {
            // nothing to do, readKeysAndValues swallowed the NONE.
        }

        private ArrayValue StartArray()
        {
            sbyte type = ( sbyte ) buf.GetByte();

            XType customStructType;
            if(type == TypeCode.CUSTOM || type == TypeCode.STRUCT)
                customStructType = ReadType();
            else
                customStructType = null;

      //      int dim = ( int ) ReadValue( intValidator );
            int dim = ReadIntegerValue().Value;

            int length = ReadIntegerValue().Value;

            Object array = AllocateArrayValue( type, customStructType, dim, length );

            return new ArrayValue( array, type, customStructType, dim );
        }

        public void EndArray( ArrayValue array )
        {
            array.Compact();
        }

        private XType ReadType()
        {
       //     int? id = ( int? ) ReadValue( intValidator );
            int? id = ReadIntegerValue();
            XType t = vf.GetType(id);
            if (t == null)
                throw new ArgumentException("unknown type " + id);
            return t;
        }

        private int? ReadIntegerValue() 
	    {
		//    return (int?)ReadValue( intValidator );
           return Convert.ToInt32(ReadValue(intValidator));
	    }

        private readonly Validator intValidator = Validator_int.Get( 0 );

        ///////////////////////////
        // LOCAL UTILITY METHODS //
        ///////////////////////////

        private Object ValidateValue( Validator v, Boolean noneOk, Object value )
        {
            if ( v == null )
                throw new Exception( "No validator" );

            if ( value == null )
                return null;

            if (noneOk && value == NONE)
                return value;

            return v.ValidateValue(value);
        }

        private Object ReadValue( Validator v )
        {
            return ReadValue( v, false );
        }

        private Object ReadValue( Validator v, Boolean noneOk )
        {
            sbyte type = ( sbyte ) buf.GetByte();

            switch ( type )
            {
                case TypeCode.NULL:
                    return ValidateValue( v, false, null );

                case TypeCode.NONE:
                    return ValidateValue(v, noneOk, NONE);

                case TypeCode.BOOLEAN_FALSE:
                    return ValidateValue( v, false, false );

                case TypeCode.BOOLEAN_TRUE:
                    return ValidateValue( v, false, true );

                case TypeCode.BYTE:
                    return  ValidateValue( v, false, buf.GetByte() );

                case TypeCode.SHORT:
                    return  ValidateValue( v, false,  buf.GetShort() );

                case TypeCode.INT:
                    return  ValidateValue( v, false, buf.GetInt() );

              

                case TypeCode.LONG:
                    return ValidateValue( v, false, buf.GetLong() );

                case TypeCode.FLOAT:
                    return ValidateValue( v, false, buf.GetFloat() );

                case TypeCode.DOUBLE:
                    return ValidateValue( v, false, buf.GetDouble() );

                case TypeCode.BYTES:
                    return ValidateValue( v, false, ReadSbytes() );

                case TypeCode.EMPTY_STRING:
                    return ValidateValue( v, false, "" );

                case TypeCode.STRING:
                    //return new String( sArray, 0, sArray.Length, Encoding.UTF8 );
                    //return Encoding.UTF8.GetString( ReadBytes() );
                    return ValidateValue( v, false, Encoding.UTF8.GetString( ReadBytes() ) );

                //case TypeCode.BOOLS:
                //case TypeCode.SHORTS:
                //case TypeCode.INTS:
                //case TypeCode.LONGS:
                //case TypeCode.FLOATS:
                //case TypeCode.DOUBLES:
                case TypeCode.ARRAY:
                    return ValidateValue( v, false, FromArrayValue( ReadArray( v ) ) );

                case TypeCode.STRUCT:
                case TypeCode.CUSTOM:
                    return ValidateValue( v, false, vf.ImportCustomValue( ReadStruct() ) );

                default:
                    if (type >= TypeCode.MIN_TINY_INT && type <= TypeCode.MAX_TINY_INT)
                        return ValidateValue(v, false, type);
                    throw new Exception( "unsupported type code "+type );

            }

        }



        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        /// Exception:
        ///     throws Exception
        private byte[] ReadBytes()
        {
            int n = ReadIntegerValue().Value;
            byte[] b = new byte[ n ];
            buf.GetFully( b );
            return b;
        }

        /// <summary>
        /// Reads a Message from the data buffer.
        /// </summary>
        /// <param name="vf">the value factory.</param>
        /// <param name="buf">the data buffer.</param>
        /// <returns>the message read.</returns>
        /// Exception:
        ///     throws Exception
        public static Message ReadMessage( ValueFactory vf, byte[] buf )
        {
            return ReadMessage( vf, new FlexBuffer( buf ));
        }


        /// <summary>
        /// Reads a Message from the data buffer.
        /// </summary>
        /// <param name="vf">the value factory for the message</param>
        /// <param name="buf">the data buffer</param>
        /// <param name="off">the offset to start of message</param>
        /// <param name="len">the length of the message</param>
        /// <returns>the message read</returns>
        /// Exception:
        ///     throws Exception
        public static Message ReadMessage( ValueFactory vf, byte[] buf, int off,
            int len )
        {
            return ReadMessage( vf, new FlexBuffer( buf, off, len ));
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="vf"></param>
        /// <returns>the message read</returns>
        /// Exception:
        ///     throws Exception
        public static Message ReadMessage( ValueFactory vf, FlexBuffer buf)
        {
            BinaryTaggedDataInput tdi = new BinaryTaggedDataInput( vf);
            return tdi.ReadMessage(buf);
        }


        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        /// Exception:
        ///     throws Exception
        private sbyte[] ReadSbytes()
        {
            int n = ( int ) ReadValue( intValidator );
            sbyte[] b = new sbyte[n];

            buf.GetFully( (byte[])(Array)b );

            
            
        /*   sbyte[] sb = new sbyte[ n ];
            Buffer.BlockCopy( b, 0, sb, 0, Buffer.ByteLength( b ) ); */
    //      sbyte[] sb = (sbyte[])(Array)b;


            return b;
        }

    }
}
